lms/djangoapps/shoppingcart/views.py

```diff
@@ -401,6 +400,26 @@ def register_code_redemption(request, registration_code):
         return render_to_response(template_to_render, context)
 
 
+def _enrollment_is_upgradeable_with_code(course, user, redemption_code):
+    """Checks to see if the user's enrollment can be updated by the code.
+
+    Check to see if the enrollment code and the user's enrollment match. If they are different, the code
+    may be used to alter the enrollment of the user. If the user is inactive, will return True, since
+    the user may use the code to re-activate an enrollment as well.
+
+    Args:
+        course (CourseDescriptor): The course to check for enrollment.
+        user (User): The user that will be using the redemption code.
+        redemption_code (CourseRegistrationCode): The redemption code that will be used to update the user's enrollment.
+
+    Returns:
+        True if the redemption code can be used to upgrade the enrollment, or re-activate it.
+
+    """
+    enrollment_mode, is_active = CourseEnrollment.enrollment_mode_for_user(user, course.id)
+    return not is_active or enrollment_mode != redemption_code.mode_slug
```

Comment:
According to its docstring, this function is used to determine if an enrollment code can be used to *update* a user's enrollment, not just upgrade it. The Boolean returned by the function also indicates whether a user's enrollment is eligible for an update. In light of that, I think it would be a good idea to rename this function `_enrollment_is_updateable_with_code`.

Also, to make it clear why a mode mismatch means an upgrade is possible, it might be worth adding a comment here noting that "A redemption code must be associated with a paid mode."